### PR TITLE
PagerDuty reactor

### PR DESCRIPTION
**Pull request by mmcclimon**
NOTE: this is almost certainly incompatible with the VO reactor. It's possible they might both work at the same time, but I haven't tested it.

This is more or less a direct port of the VictorOps reactor, with the necessary changes for the differences between the two services. It's likely we'll want to improve this somewhat, but this gets us to feature parity. The commit messages have more detail, where necessary.

The most notable thing I _didn't_ port was extra behavior on demaint: that is `demaint /resolve` is no longer a thing, nor is `demaint /noresolve`, which never quite worked right anyway.

### CODE REVIEW

**Comment by Davis-A** [comment]
Sorry i didn't get to this today.  Took most of the day off.  I'll take a look tomorrow :)

File: lib/Synergy/Reactor/PagerDuty.pm [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+    ->then(sub (@maint) {
+      unless (@maint) {
+        return Future->fail('not in maint');
+      }
+
+      $self->last_maint_warning_time($current_time);
+      $self->save_state;
+
+      # we'll warn if any of them is older than us
+      # maint startedAt is unix time * 1000
```
Comment (Davis-A) [inline lib/Synergy/Reactor/PagerDuty.pm:390]: Comment not applicable for PD

**Review by Davis-A — CHANGES_REQUESTED** [review]

File: lib/Synergy/Reactor/PagerDuty.pm [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+  my $sid = $self->service_id;
+
+  my $whose = $arg->{whose};
+  Carp::confess("_resolve_incidents called with bogus args")
+    unless $whose && ($whose eq 'all' || $whose eq 'own');
+
+  my $only_acked = $arg->{only_acked} // ($whose eq 'own' ? 1 : 0);
+
+  # XXX pagination?
+  return $self->_pd_request(GET => "/incidents?service_ids[]=$sid&statuses[]=triggered&statuses[]=acknowledged&limit=100")
```
[the remark this answers is not in the document]
Comment (mmcclimon) [reply lib/Synergy/Reactor/PagerDuty.pm:661]: Not an idiot! Those multi-arg query params are weird.

[the remark this answers is not in the document]
Comment (Davis-A) [reply lib/Synergy/Reactor/PagerDuty.pm:661]: Yeah it's super weird.  I also just remembered, Butch mentioned you can 'snooze' an incident like in VO.  But i have no idea what status they'll have as the docs don't mention it.

**Comment by Davis-A** [comment]
> The most notable thing I didn't port was extra behavior on demaint: that is demaint /resolve is no longer a thing, nor is demaint /noresolve, which never quite worked right anyway.

I thought about this.  I think the default behaviour to resolve is wrong.  But i am in favour of a /resolve option.  But that can totally be added later.

FYI /noresolve was fixed in #146

**Comment by mmcclimon** [comment]
> I thought about this. I think the default behaviour to resolve is wrong. But i am in favour of a /resolve option. But that can totally be added later.

@Davis-A `demaint /resolve` isn't really useful, because PagerDuty (and now, actually, VO too) doesn't fire incidents at all during maint, so there's nothing _to_ resolve. The previous behavior resolved everything that fired since start of maint plus 10 minutes, I think, which was also very confusing IMO.

**Comment by Davis-A** [comment]
Oh ... i was under the impression the alerts still fired they just weren't displayed by the integration any more.  That is ... most annoying.

File: lib/Synergy/Reactor/PagerDuty.pm (hunk 1 above)
Comment (mmcclimon) [reply lib/Synergy/Reactor/PagerDuty.pm]: Good catch, thanks!

File: lib/Synergy/Reactor/PagerDuty.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmcclimon) [reply lib/Synergy/Reactor/PagerDuty.pm:659]: https://support.pagerduty.com/docs/editing-incidents#snooze-an-incident implies that snoozing is only available for acknowledged incidents, which makes sense.

File: lib/Synergy/Reactor/PagerDuty.pm [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+    })
+    ->else(sub {
+      $event->reply("Something went wrong acking incidents. Sorry!");
+    })
+    ->retain;
+}
+
+sub _ack_all ($self, $event) {
+  my $sid = $self->service_id;
+  # XXX is this limit sufficient?
```
[the remark this answers is not in the document]
Comment (mmcclimon) [reply lib/Synergy/Reactor/PagerDuty.pm]: @Davis-A I have added two more commits to deal with this in a more general way, which should be sufficient. Let me know if it's not!

**Review by Davis-A — APPROVED** [review]

**Comment by Davis-A** [comment]
This looks fantastic.  Thanks for doing this Michael!

**Pull request merged**